std/traits.d

```diff
@@ -6839,3 +6839,348 @@ unittest
+        alias FA = FunctionAttribute;
+        enum isPropertyFunction = (fa!(FunctionTypeOf!(X[0])) & FA.property) != 0;
+    }
+    else
+        enum isPropertyFunction = false;
+}
+
+///
+unittest
+{
+    static void func(){}
+    static assert(!isPropertyFunction!func);
+
+    struct S
+    {
+        void func() @property {}
+    }
+    static assert(isPropertyFunction!(S.func));
+}
+
+unittest
+{
+    // for (nested) (property) function symbols
+    static void func(){}
+    static @property int prop(){ return 0; }
+    void nested_func(){}
+    @property int nested_prop(){ return 0; }
+    static assert( isFunction!func);
+    static assert( isFunction!prop);
+    static assert( isFunction!nested_func);
+    static assert( isFunction!nested_prop);
+    static assert(!isPropertyFunction!func);
+    static assert( isPropertyFunction!prop);
+    static assert(!isPropertyFunction!nested_func);
+    static assert( isPropertyFunction!nested_prop);
+
+    // for functino types
+    alias F = typeof(func);
+    alias P = SetFunctionAttributes!(F, "D", FunctionAttribute.property);
+    alias DG = typeof(&nested_func);
+    alias DP = typeof(&nested_prop);
+    static assert( isFunction!F);
+    static assert( isFunction!P);
+    static assert(!isFunction!DG);  // DG is delegate type, not function type
+    static assert(!isFunction!DP);  // DP is delegate type, not function type
+    static assert(!isPropertyFunction!F);
+    static assert( isPropertyFunction!P);
+    static assert(!isPropertyFunction!DG);
+    static assert(!isPropertyFunction!DP);
+
+    // other cases
+    int n;
+    struct S {}
+    static assert(!isFunction!int);
+    static assert(!isFunction!void);
+    static assert(!isFunction!n);
+    static assert(!isFunction!S);
+    static assert(!isFunction!std);
+    static assert(!isPropertyFunction!int);
+    static assert(!isPropertyFunction!void);
+    static assert(!isPropertyFunction!n);
+    static assert(!isPropertyFunction!S);
+    static assert(!isPropertyFunction!std);
+
+    enum E { x = 10 }
+    static assert(!isFunction!(E.x));
+    static assert(!isPropertyFunction!(E.x));
+}
+
+/**
+ * Detect whether symbol $(D X) is a run-time variable, i.e. has an address and
+ * a value and is not an `enum`.
+ *
+ * Returns:
+ *     `bool`
+ */
+template isVariable(X...) if (X.length == 1)
+{
+    static if (!is(X[0]) &&
+               !is(typeof(X[0]) == void) &&
+               !isFunction!(X[0]))
+    {
+        enum isVariable =
+            is(typeof({ auto ptr = &X[0]; }))
+         || is(typeof({ enum off = X[0].offsetof; }));
+    }
+    else
+        enum isVariable = false;
+}
+
+///
+unittest
+{
+    int my;
+    immutable int iy = 0;
+    enum int ey = 0;
+    static assert( isVariable!my);
+    static assert( isVariable!iy);
+    static assert(!isVariable!ey);
+}
+
+/**
+ * Detect whether $(D X) is an enum type, or manifest constant.
+ *
+ * Returns:
+ *     `bool`
+ */
+template isEnum(X...) if (X.length == 1)
```

Comment:
I don't think treating enum types the same as manifest constants is a useful trait. Is there any justification for this?